experiment/simulation/js/validator.js

```diff
@@ -0,0 +1,229 @@
+        result.className = "success-message";
+    }
+    else {
+        result.innerHTML = "<span>&#10007;</span> Fail";
+        result.className = "failure-message";
+    }
+}
+
+// JK Flip FLop Tester
+export function testJKFF(inputJ, inputK, inputClk, outputQB, outputQA)  // This function takes 4 ids of the respective Gates
+{
+    let gates_list = gates;
+    let flipflops_list = flipFlops;
+    let j = gates_list[inputJ];
+    let k = gates_list[inputK];
+    let clk = gates_list[inputClk];
+    let circuitIsCorrect = true;
+
+
+    let qa = gates_list[outputQA];
+    let qb = gates_list[outputQB];
+    // each list element consists of 5 values, j,k,Clk,Q,Qbar
+    const expectedOutputs = [[false, false], [false, true], [true, false], [true, true]];
+    j.setOutput(true);
+    k.setOutput(true);
+    clk.setOutput(false);
+    if (testSimulation(gates_list, flipflops_list)) {
+        clk.setOutput(true);
+        testSimulation(gates_list, flipflops_list);
+        clk.setOutput(false);
+        testSimulation(gates_list, flipflops_list);
+        let first_output = -1;
+        for (let i = 0; i < 4; i++) {
+            if (qa.output == expectedOutputs[i][1] && qb.output == expectedOutputs[i][0]) {
+                first_output = i;
+                break;
+            }
+        };
+        if (first_output == -1) {
+            circuitIsCorrect = false;
+        }
+        else {
+            for (let i = 1; i < 4; i++) {
+                clk.setOutput(true);
+                testSimulation(gates_list, flipflops_list);
+                clk.setOutput(false);
+                testSimulation(gates_list, flipflops_list);
+                // check if output is correct
+                if ((qb.output !== expectedOutputs[(i + first_output) % 4][0] || qa.output !== expectedOutputs[(i + first_output) % 4][1])) {
+                    circuitIsCorrect = false;
+                }
+            }
+        }
+        const result = document.getElementById('result');
+
+        if (circuitIsCorrect) {
+            result.innerHTML = "<span>&#10003;</span> Success";
+            result.className = "success-message";
+        }
+        else {
+            result.innerHTML = "<span>&#10007;</span> Fail";
+            result.className = "failure-message";
+        }
+    }
+}
+
+
+// D Flip FLop Tester
+export function testDFF(inputOri, inputClk, outputA, outputB, outputC)  // This function takes 4 ids of the respective Gates
+{
+    let gates_list = gates;
+    let flipflops_list = flipFlops;
+    let ori = gates_list[inputOri];
+    let clk = gates_list[inputClk];
+    let outputa = gates_list[outputA];
+    let outputb = gates_list[outputB];
+    let outputc = gates_list[outputC];
+    let circuitIsCorrect = true;
+    //ori.setOutput(false);
+    clk.setOutput(false);
+
+    // each list element consists of 4 values, D,Clk,Q,Qbar
+    const expectedOutputs = [[false, true, false], [true, false, false], [false, false, true]]

```

Comment:
Check more than once

Reply:
now we are checking for 2*n possibilities where n is the number of unique states.